src/routes/activityRouter.js

```diff
@@ -0,0 +1,19 @@
+const express = require("express");
+const router = express.Router();
+
+const {
+  getAllActivities,
+  saveActivity,
+  updateActivityVote,
+} = require("../controllers/activityController");
+
+// @route   GET /api/v1/activity
+router.get("/", getAllActivities);
+
+// @route   POST /api/v1/activity/create
+router.post("/create", saveActivity);
+
+// @route   PUT /api/v1/activity/vote/:_id
+router.put("/vote/:_id", updateActivityVote);
```

Comment:
To follow API route design conventions, and since the resource being updated is an `activity`, we shouldn't include `vote` in the route. This structure makes it look like `vote` is a (nested) resource being updated (and that the id param is the id of a `vote`), but actually `vote` is a field on `activity`. Info about what fields are being updated should be passed in the body of the request. Does that make sense?

Reply:
Yes, this does make sense. However, the only reason I put vote in the URI is because this specific endpoint is exclusively responsible for updating the vote count field of the model. And in the future, there can be other endpoints for updating either the activity resource as a whole, or some other part of it. So I am not sure if there is a better solution for this for now.

Reply:
Whenever possible, we should use the [standard methods](https://cloud.google.com/apis/design/standard_methods) for resources of List, Get, Create, Update, and Delete. 

So for each resource (often resource = model/table/collection, but not necessarily), we should almost always only have routes like...
- `GET /activities` (List)
- `GET /activities/:id` (Get)
- `POST /activities` (Create)
- `PUT or PATCH /activities/:id` (Update)
- `DELETE` /activities/:id` (Delete)

If we were allowing the client to update the vote count directly (like in the first implementation), I'd say we definitely want to use the standard PUT update method (even though we're only updating one field on the `activity` record. We could change the route to allow more `activity` fields to be updated in the future). But the new implementation (allowing the client to only increment or decrement an existing count) is a little less clear, since an action is being performed, rather than a straight update of data based on the values passed. Still, I think it would be good for us to try to map it to a standard method for some resource.

What will the route look like once users are casting votes for an activity? Maybe it would be a post request to create a new vote?

`POST /votes`

What do you think about making that the route now?

Reply:
In the current implementation of the `activity` schema, a `votes` field which is set to the value of 0 is added by default upon creating a new `activity` resource. And the way the `votes` is getting changed right now is through a `PUT` request. So I think it would be ok if we are only updating a single field of an `activity` resource through the `PUT` request for now, and we can change this to include other fields to be updated as well in the future.

And I am not sure about the idea of creating a separate `POST` request for creating a new vote. Since in the current logic a vote is not getting created, rather its getting updated.

I will update the `PUT` request's URI in order to follow the standard method.

Reply:
I'm definitely not as up to speed on these topics as  Angel, but I think maybe the reason we would want to create votes as new entries in a table would be when we want to track "Darga voted for activity 1234 at 12:14pm 1/2/34"

Then later when we try to display the full list of votes for a given option we can see the list of all votes with all that information?

Reply:
> So I think it would be ok if we are only updating a single field of an activity resource through the PUT request for now, and we can change this to include other fields to be updated as well in the future.

Okay, I'm good with this. 👍🏻

> And I am not sure about the idea of creating a separate POST request for creating a new vote. Since in the current logic a vote is not getting created, rather its getting updated.

So this is pretty nuanced, but when designing an API, we decide what `resources` to make available for CRUD operations. Often, this lines up directly with what we call models/tables/collections on the backend, but not always. In this case, even though a vote is currently a field on an activity, we could decide that for the purpose of the API design, it's a resource. (We are actually creating a vote. The client doesn't have to know the structure of our data.) And as Darga mentioned, a vote probably should be an independent model/collection in future versions when we start associating votes with users anyway. So we could choose to make a route `POST /votes` that updates a field on an activity now, and we can change it to creating a `vote` row later. **But your decision to make the route `PUT /activities/:id` is a good one and I'm definitely in support of that!** 👍🏻

Reply:
Yes, I understood this quite recently. And just as I saw the notification of your comment pop up, I was working on separating the votes to its own model. 😄

Having a separate model for votes is better since we'll have to reference users as well into each of the votes in a later version.

Reply:
We can do this in a later ticket though if you'd like! The frontend team will be working on submitting votes without a user attached right now, so they'll need an endpoint that allows that.

Reply:
The current implementation is all without the user model, so they'll be able to use the current routes no problem.